akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,237 @@
+import akka.stream.impl.{ Buffer => BufferImpl, PartitionedBuffer }
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.stage._
+import akka.util.OptionVal
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+private[akka] object MapAsyncPartitioned {
+  def apply[In, Out, Partition](
+      parallelism: Int,
+      perPartition: Int,
+      partitioner: In => Partition,
+      f: (In, Partition) => Future[Out]): GraphStage[FlowShape[In, Out]] =
+    if (parallelism < 1) throw new IllegalArgumentException("parallelism must be at least 1")
+    else if (perPartition < 1) throw new IllegalArgumentException("perPartition must be at least 1")
+    else if (perPartition >= parallelism) {
+      // no point controlling per-partition, so just use a MapAsync, which has somewhat less overhead
+      val fin = { (x: In) =>
+        f(x, partitioner(x))
+      }
+      MapAsync(parallelism, fin)
+    } else new MapAsyncPartitioned(parallelism, perPartition, partitioner, f) {}
+
+  final class Holder[P, I, O](
+      var incoming: I,
+      var outgoing: Try[O],
+      val partition: P,
+      val cb: AsyncCallback[Holder[P, I, O]])
+      extends (Try[O] => Unit) {
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive =
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) => d
+        case _ =>
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
+          d
+      }
+
+    def clearIncoming(): Unit = {
+      incoming = null.asInstanceOf[I]
+    }
+
+    def setOutgoing(t: Try[O]): Unit = {
+      outgoing = t
+    }
+
+    override def apply(t: Try[O]): Unit = {
+      setOutgoing(t)
+      cb.invoke(this)
+    }
+
+    override def toString: String = s"Holder($incoming, $outgoing, $partition)"
+  }
+
+  val NotYetThere = MapAsync.NotYetThere
+}
+
+private[akka] sealed abstract case class MapAsyncPartitioned[In, Out, Partition] private (
+    parallelism: Int,
+    perPartition: Int,
+    partitioner: In => Partition,
+    f: (In, Partition) => Future[Out])
+    extends GraphStage[FlowShape[In, Out]] {
+  import MapAsyncPartitioned._
+
+  private val in = Inlet[In]("MapAsyncPartitioned.in")
+  private val out = Outlet[Out]("MapAsyncPartitioned.out")
+
+  override def initialAttributes = DefaultAttributes.mapAsync and SourceLocation.forLambda(f)
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      var buffer: PartitionedBuffer[Partition, Holder[Partition, In, Out]] = _
```

Comment:
Can you motivate the need for the custom buffer impl vs using something like `var buffer: Map[Partition, Buffer[Holder]]` a bit (the benefit vs the code to maintain is not obvious to me)?

Reply:
see line 231 (`else if (buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)`) thread.

> This need to keep the buffers in sync motivated the special buffer implementation.  An earlier implementation of this had some edge cases where the buffers got out of sync and the stage deadlocked, so implementing `PartitionedBuffer` and `ChainedBuffer` (which has the side benefit of (thanks to exposing `DynamicQueue`), for the sunny-day periods when there's at most `perPartition` elements per partition, not having to allocate approx. `parallelism / perPartition` fixed-size buffers of size `parallelism`) separately let me factor out keeping multiple buffers in sync.